src/Template/ItemCombinator.php

```diff
@@ -35,16 +36,26 @@ public static function union(array ...$itemArrays): array
             foreach ($itemArray as $item) {
                 $itemName = $item->getName();
                 if (isset($union[$itemName])) {
-                    $class = get_class($item);
-                    $union[$itemName] = new $class(
-                        $itemName,
-                        TypeCombinator::union($union[$itemName]->getType(), $item->getType())
-                    );
+                    $union[$itemName] = $item->withType(TypeCombinator::union($union[$itemName]->getType(), $item->getType()));
                 } else {
                     $union[$itemName] = $item;
                 }
             }
         }
         return array_values($union);
     }
+
+    /**
+     * @param NameTypeItem[] $items
+     * @return NameTypeItem[]
+     */
+    public static function resolveTemplateTypes(array $items, string $declaringClass, ?string $currentClass): array
```

Comment:
how is template type resolver connected with item combinator?

Reply:
Well only vaguely bit I did not find better place for it. I think af it as sort of combining Items with current template. Or would you preffer separate ItemTemplateTypeResolver? Or something like ItemHelper?

Reply:
What is this for? I don't get it. You use it in component finder, filter finder etc.

Reply:
When we collect types we might collect `T` reffering to generic template type `@template T of Something`. We then need to replace `T` with actual type passed to generic in given implementation of generic type.

Reply:
We cannor do it at collect time because when we are collecting types from generic parent it can be used in many different childs. We have to replave it in finder when we know actualClass

Reply:
what is array $items for? I get lost because of them :D

Reply:
It is for example array or Variable and this method check types of all variables in array and resolves genetic template types with actual types. So tresult is array of Variable with their types resolved.